File: Year-2023/LeetCode-February-Challenge/BinaryTreeZigzagLevelOrderTraversal.java
import java.util.*;
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    private void levelorder(TreeNode root){
        boolean flag = false;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            int size = q.size();
            LinkedList<Integer> temp = new LinkedList<>();
            flag = !flag;
            while(size-- != 0){
                TreeNode curr = q.poll();
                if(flag)
                    temp.add(curr.val);
                else temp.addFirst(curr.val);
                if(curr.left != null) q.add(curr.left);
                if(curr.right != null) q.add(curr.right);
            }
            res.add(temp);
        }
    }
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if(root == null) return new ArrayList<>();
        levelorder(root);
        return res;
    }
}
public class BinaryTreeZigzagLevelOrderTraversal{
    public static void main(String[] args) {
        
    }
}

File: Year-2023/LeetCode-February-Challenge/BinaryTreeZigzagLevelOrderTraversal.cpp
#include<bits/stdc++.h>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution {
private:
    vector<vector<int>> res;
    void levelorder(TreeNode *root){
        queue<TreeNode *> q;
        q.push(root);
        while(!q.empty()){
            int size = q.size();
            vector<int> temp;
            while(size--){
                TreeNode *curr = q.front();
                temp.push_back(curr->val);
                q.pop();
                if(curr->left) q.push(curr->left);
                if(curr->right) q.push(curr->right);
            }
            if(res.size() % 2 != 0) reverse(temp.begin(), temp.end());
            res.push_back(temp);
        }
    }
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(!root) return {};
        levelorder(root);
        return res;
    }
};
